grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java

```diff
@@ -136,6 +174,123 @@ public static GraqlTraversal createTraversal(PatternAdmin pattern, EmbeddedGrakn
+        allFragments.stream()
+                .filter(LabelFragment.class::isInstance)
+                .forEach(fragment -> {
+                    // TODO: labels() should return ONE label instead of a set
+                    Type type = tx.getType(((LabelFragment) fragment).labels().iterator().next());
+                    if (type != null && !type.isRole()) labelVarTypeMap.put(fragment.start(), type);
+                });
+        if (labelVarTypeMap.isEmpty()) return;
+
+        // find all vars with direct or indirect out isa edges
+        Multimap<Var, Type> instanceVarTypeMap = HashMultimap.create();
+        int size;
+        do {
+            size = instanceVarTypeMap.size();
+            allFragments.stream()
+                    .filter(fragment -> labelVarTypeMap.containsKey(fragment.start()))
+                    .filter(fragment -> fragment instanceof InIsaFragment || fragment instanceof InSubFragment)
+                    .forEach(fragment -> instanceVarTypeMap.put(fragment.end(), labelVarTypeMap.get(fragment.start())));
+        } while (size != instanceVarTypeMap.size());
+
+        // relationship vars and its role player vars
+        Multimap<Var, Var> relationshipRolePlayerMap = HashMultimap.create();
+        allFragments.stream().filter(OutRolePlayerFragment.class::isInstance)
+                .forEach(fragment -> relationshipRolePlayerMap.put(fragment.start(), fragment.end()));
+
+        // find all the relationships requiring type inference
+        Iterator<Var> iterator = relationshipRolePlayerMap.keySet().iterator();
+        while (iterator.hasNext()) {
+            Var relationship = iterator.next();
+
+            // the relation should have at least 2 known role players so we can infer something useful
+            if (instanceVarTypeMap.containsKey(relationship) ||
+                    relationshipRolePlayerMap.get(relationship).size() < 2) {
+                iterator.remove();
+            } else {
+                final int[] numRolePlayersHaveType = {0};
+                relationshipRolePlayerMap.get(relationship).forEach(
+                        rolePlayer -> {
+                            if (instanceVarTypeMap.containsKey(rolePlayer)) {
+                                numRolePlayersHaveType[0]++;
+                            }
+                        });
+                if (numRolePlayersHaveType[0] < 2) {
+                    iterator.remove();
+                }
+            }
+        }
+        if (relationshipRolePlayerMap.isEmpty()) return;
+
+        // for each type, get all possible relationship type it could be in
+        Map<Type, Set<RelationshipType>> relationshipMap = new HashMap<>();
+        getAllPossibleRelationships(relationshipMap, tx.admin().getMetaEntityType());
+        getAllPossibleRelationships(relationshipMap, tx.admin().getMetaAttributeType());
+        getAllPossibleRelationships(relationshipMap, tx.admin().getMetaRelationType());
```

Comment:
Isn't doing these potentially expensive for bigger ontologies? Could we call `getAllPossibleRelationships` on a specific type when we need it in `getAllPossibleRelationShipTypesOfARolePlayer`?